apps/src/p5lab/spritelab/commands/criterionCommands.js

```diff
@@ -583,9 +583,92 @@ export const commands = {
     return result;
   },
 
-  // Returns true if the student set a variable to any number, string, or boolean.
-  variableCreated() {
-    const result = this.studentVars.length >= 1;
-    return result;
+  // Returns true if a minimum number of foreground effects are currently active.
+  // Disregards their initial render frame.
+  foregroundEffectsActive(min = 1) {
+    return this.getForegroundEffects().length >= min;
+  },
+
+  // Returns true if a minimum number of foreground effects were first rendered
+  // this frame (e.g. using an event block)
+  foregroundEffectRenderedThisFrame(min = 1) {
+    const effects = this.getForegroundEffects();
+    const newEffects = effects.filter(
+      effect => effect.renderFrame === this.currentFrame()
+    );
+    return newEffects.length >= min;
+  },
+
+  // Returns true if a minimum number of variable bubbles are currently being drawn.
+  variableBubblesCreated(min = 1) {
+    return this.getVariableBubbles().length >= min;
+  },
+
+  // Returns true if a minimum number of watched variables have a valid value.
+  // If unset, a variable's value is an empty string.
+  variableValueSet(min = 1) {
+    const studentVars = this.getVariableBubbles();
+    const filteredVars = studentVars.filter(
+      studentVar => this.getVariableValue(studentVar.name) !== ''
+    );
+    return filteredVars.length >= min;
+  },
+
+  // Returns true if a watched variable's value changed this frame.
+  // Optionally checks if the variable values have increased or decreased.
+  // Since this command requires a variable bubble, it should be used in conjunction
+  // with variableBubblesCreated.
+  variableValueChanged(changeType = '') {
```

Comment:
is it worth documenting (in the comment above?) what the valid values for `changeType` are? I'm not sure how we generally document these functions

Reply:
Yeah I think that’s worth adding. Good idea! As far documentation goes, right now it’s really just the comments in this file. I’d love to eventually publish something easier to read for levelbuilder.